# DecodeLabs Data Analytics Internship

## Project 1: Data Cleaning & Preparation

### Objective
The objective of this project is to clean and prepare a raw e-commerce dataset by identifying missing values, checking duplicate records, correcting data formats, and validating data consistency to ensure high-quality data for further analysis.

### Tools Used
- Python
- Pandas
- NumPy
- Jupyter Notebook

## 1. Importing Required Libraries

In this step, the necessary Python libraries are imported for data manipulation and analysis.

In [1]:
import pandas as pd
import numpy as np

## 2. Loading the Dataset

The dataset is loaded into a Pandas DataFrame to begin the data cleaning process.

In [26]:
df = pd.read_excel("Dataset for Data Analytics.xlsx")
print("Dataset Shape:", df.shape)

Dataset Shape: (1200, 14)


In [4]:
df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


## 3. Dataset Overview

This section examines the dataset structure, number of records, columns, and data types.

In [5]:
print("Rows and Columns:")
print(df.shape)

print("\nColumn Names:")
print(df.columns)

print("\nData Types:")
print(df.dtypes)

Rows and Columns:
(1200, 14)

Column Names:
Index(['OrderID', 'Date', 'CustomerID', 'Product', 'Quantity', 'UnitPrice',
       'ShippingAddress', 'PaymentMethod', 'OrderStatus', 'TrackingNumber',
       'ItemsInCart', 'CouponCode', 'ReferralSource', 'TotalPrice'],
      dtype='str')

Data Types:
OrderID                       str
Date               datetime64[us]
CustomerID                    str
Product                       str
Quantity                    int64
UnitPrice                 float64
ShippingAddress               str
PaymentMethod                 str
OrderStatus                   str
TrackingNumber                str
ItemsInCart                 int64
CouponCode                    str
ReferralSource                str
TotalPrice                float64
dtype: object


In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   str           
 1   Date             1200 non-null   datetime64[us]
 2   CustomerID       1200 non-null   str           
 3   Product          1200 non-null   str           
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   str           
 7   PaymentMethod    1200 non-null   str           
 8   OrderStatus      1200 non-null   str           
 9   TrackingNumber   1200 non-null   str           
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       891 non-null    str           
 12  ReferralSource   1200 non-null   str           
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[us](1), float64(2), int64(2), str(9)


### Observation

The dataset contains 1200 records and 14 columns. Most columns contain complete data, while the CouponCode column contains missing values that require cleaning.

## 4. Missing Value Analysis

The dataset is checked for null or missing values to identify incomplete information.

In [7]:
df.isnull().sum()

OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64

### Observation

The CouponCode column contains 309 missing values. All other columns have complete information.

In [8]:
missing_values = df.isnull().sum()

missing_values[missing_values > 0]

CouponCode    309
dtype: int64

In [9]:
missing_percentage = (
    df.isnull().sum() / len(df)
) * 100

missing_percentage[missing_percentage > 0]

CouponCode    25.75
dtype: float64

## 5. Handling Missing Values

Missing CouponCode values are replaced with "No Coupon" because a blank value indicates that no promotional coupon was used during the purchase.

In [10]:
df["CouponCode"].value_counts(dropna=False)

CouponCode
FREESHIP    313
NaN         309
WINTER15    292
SAVE10      286
Name: count, dtype: int64

In [11]:
df["CouponCode"] = df["CouponCode"].fillna("No Coupon")

In [12]:
df.isnull().sum()

OrderID            0
Date               0
CustomerID         0
Product            0
Quantity           0
UnitPrice          0
ShippingAddress    0
PaymentMethod      0
OrderStatus        0
TrackingNumber     0
ItemsInCart        0
CouponCode         0
ReferralSource     0
TotalPrice         0
dtype: int64

### Observation

All missing values have been successfully handled and the dataset now contains zero null values.

## 6. Duplicate Record Detection

The dataset is checked for duplicate records and duplicate Order IDs to ensure uniqueness and data integrity.

In [13]:
df.duplicated().sum()

np.int64(0)

In [14]:
df["OrderID"].duplicated().sum()

np.int64(0)

### Observation

No duplicate rows and no duplicate Order IDs were found in the dataset.

## 7. Date Format Standardization

Dates are converted into a consistent YYYY-MM-DD format for improved readability and compatibility with analytical tools.

In [15]:
df["Date"].head()

0   2023-01-04
1   2024-08-23
2   2024-02-27
3   2023-10-15
4   2025-05-08
Name: Date, dtype: datetime64[us]

In [16]:
df["Date"].dtype

dtype('<M8[us]')

In [17]:
df["Date"] = pd.to_datetime(df["Date"])

In [18]:
df["Date"] = df["Date"].dt.strftime("%Y-%m-%d")

In [19]:
df["Date"].head()

0    2023-01-04
1    2024-08-23
2    2024-02-27
3    2023-10-15
4    2025-05-08
Name: Date, dtype: str

### Observation

All dates were successfully standardized into a uniform format.

## 8. Text Data Cleaning

Text columns are cleaned by removing unnecessary spaces and standardizing capitalization.

In [22]:
text_columns = [
    "Product",
    "PaymentMethod",
    "OrderStatus",
    "ReferralSource",
    "CouponCode"
]

for col in text_columns:
    df[col] = df[col].astype(str).str.strip().str.title()

### Observation

Text fields are now consistently formatted, reducing the possibility of duplicate categories caused by inconsistent spelling or capitalization.

## 9. Business Rule Validation

The TotalPrice column is validated by comparing it against Quantity × UnitPrice.

In [23]:
calculated_total = (
    df["Quantity"] *
    df["UnitPrice"]
).round(2)

mismatch = df[
    calculated_total !=
    df["TotalPrice"].round(2)
]

len(mismatch)

0

### Observation

No inconsistencies were found between the calculated and recorded total prices.

## 10. Final Verification

A final quality check is performed to confirm that the dataset is clean and ready for analysis.

In [24]:
print("Missing Values:", df.isnull().sum().sum())
print("Duplicate Rows:", df.duplicated().sum())
print("Duplicate Order IDs:", df["OrderID"].duplicated().sum())

Missing Values: 0
Duplicate Rows: 0
Duplicate Order IDs: 0


### Observation

The final dataset contains:
- 0 Missing Values
- 0 Duplicate Rows
- 0 Duplicate Order IDs

The dataset is fully cleaned and prepared for analysis.

## 11. Exporting the Cleaned Dataset

The cleaned dataset is saved for future analysis and reporting.

In [28]:
df.to_excel("cleaned_dataset.xlsx", index=False)

## Data Quality Summary

| Check | Result |
|---------|---------|
| Missing Values | Resolved |
| Duplicate Rows | None Found |
| Duplicate Order IDs | None Found |
| Date Formatting Issues | Corrected |
| Text Formatting Issues | Corrected |
| TotalPrice Validation | Verified |

The dataset successfully passed all quality checks and is ready for further analysis.

# Conclusion

The dataset was successfully cleaned and prepared by handling missing values, verifying duplicate records, standardizing date formats, cleaning textual data, and validating numerical consistency. The final dataset is accurate, complete, and ready for further analytical tasks.